02-KernelModules_CharacterDevice/main.c

```diff
@@ -0,0 +1,259 @@
+static int proc_read(struct file *filp, char *buffer, size_t len, loff_t *offset ){
+	
+
+	int result;
+	char msg[124];
+	
+	sprintf(msg, "%d %d\n", buffer_size+buffer_increase, buffer_size+buffer_increase-used_memory);
+
+	
+	if(*offset >= strlen(msg)) {
+		*offset=0;
+		return 0;
+	}
+	
+	if(len > strlen(msg) - *offset)
+		len = strlen(msg) - *offset;
+
+	result = raw_copy_to_user((void*)buffer, msg - *offset, len);
+	
+	*offset += len;
+	
+	return len;
+}
+
+//===/dev file operations===
+
+int dev_open(struct inode *inode, struct file *filp){	
+	if(is_open){
+		printk(KERN_ERR "ChD: Failed to open - device is busy");
+		return -EBUSY;
+	}
+	is_open = 1;
+	return 0;
+}
+
+int dev_release(struct inode *inode, struct file *filp){
+	is_open = 0;
+	return 0;
+}
+
+ssize_t dev_read(struct file *filep, char *buffer, size_t len, loff_t *offset){
+	ssize_t retval = 0;
+
+	int remain = buffer_size - (int) (*offset);
+	if (remain == 0) {
+		goto out;
+	}
+
+	if (len > remain) {
+		len = remain;
+	}
+
+	if (raw_copy_to_user(buffer, data_buff + *offset, len)) {
+		retval = -EFAULT;
+		goto out;
+	}
+	*offset += len;
+	retval = len;
+
+	out:
+		return retval;
+}
+
+ssize_t dev_write(struct file *filep, const char *buffer, size_t len, loff_t *offset){
+
+	int error = 0;
+	int remain = buffer_size - (int) (*offset);
+	
+	used_memory += len;
+	
+	if(0 == remain)
+		return 0;
+	if(used_memory > remain)
+		return -EIO;
+	
+	error = raw_copy_from_user(data_buff + *offset, buffer, len);
+	
+	if(error){
+		printk(KERN_INFO "ChD: Failed to receive %zu characters from the user\n", len);
+		return -EFAULT;
+	}
+	else{
+		printk(KERN_INFO "ChD: Received %zu characters from the user\n", len);
+		offset += len;
+		return len;
+
+	}
+}
+
+static struct file_operations fops = {
+	.open = dev_open,
+	.read = dev_read,
+	.write = dev_write,
+	.release = dev_release,
+};
+
+//===init/exit functions===
+
+static int __init ChD_init(void){
+	
+	printk(KERN_INFO "Loading module...");
+
+	major = register_chrdev(0, DEVICE_NAME, &fops);
+	
+	if(major < 0){
+		printk(KERN_ALERT "ChD: Failed to register major number\n");
+		return major;
+	}
+	else{
+		printk(KERN_INFO "ChD: Registered major number %d\n", major);
+	}
+	
+	chdClass = class_create(THIS_MODULE, CLASS_NAME);
+
+	chDevice = device_create(chdClass, NULL, MKDEV(major, 0), NULL, DEVICE_NAME);
+
+	printk(KERN_INFO "ChD: Device initialized succesfully\n");
+
+	buffer_create();
+	
+	create_proc();
+	
+	chd_kobject = kobject_create_and_add(MODULE_NAME, kernel_kobj);
+	
+	sysfs_create_file(chd_kobject, &sysfs_attribute.attr);
+
+	return 0;
+
+}
+
+static void __exit ChD_exit(void){
+	kobject_put(chd_kobject);
+	
+	cleanup_proc();
+	
+	buffer_clean();
+	
+	device_destroy(chdClass, MKDEV(major,0));
+	
+	class_unregister(chdClass);
+	
+	class_destroy(chdClass);
+	
+	unregister_chrdev(major, DEVICE_NAME);
+	
+	printk(KERN_INFO "ChD: Module unloaded\n");
+}
+
+module_init(ChD_init);
+module_exit(ChD_exit);
+
+MODULE_LICENSE("GPL");
```

Comment:
[02-KernelModules_CharacterDevice/main.c](https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/106/files#diff-389fcb28ab8b9bd0ef8ac68f9a3f772e)

```suggestion
MODULE_LICENSE("GPL");
MODULE_AUTHOR("Name@gmail.com");
MODULE_DESCRIPTION("Modul for test Character Decices");
MODULE_VERSION("1.0")
```
Need fix description of module

Reply:
main.c it is not name of driver or module!

Need fix it

Reply:
Please checkout this pull request: https://github.com/Kernel-GL-HRK/gl-kernel-basecamp-2019/pull/105